addrelease: Handle disc number tags with totaldisc information

mineo opened the pull request:
This also removes some Python 2 compatibility code in `test.py` and adds doctests for the disc number extraction.

I don't feel strongly about using doctests, but it seemed the easiest solution without always importing unittest, even if just importing the plugin in Picard.

mineo committed:
addrelease: Split the disc number extraction into a function

It's a large enough block that it's worth it.

mineo committed:
Handle disc number tags with totaldisc information

The files of
https://musicbrainz.org/release/49b267b5-8bca-4a35-b2ce-d3cd5591c207 for example
have discnumber tags containing "1/1".

mineo committed:
Add doctests for extract_discnumber

mineo committed:
Remove Python 2 compatibility code

mineo committed:
Run doctests in test.py

mineo committed:
Install picard on travis

This allows importing picard modules in tests

mineo commented:
Well, the travis test doesn't pass (yet) because the changes from https://github.com/metabrainz/picard/pull/998 haven't been released yet. But now we even have tests for our source distributions ;-) Anyways, this can wait a while.

phw commented:
@mineo The build errors now are related to https://github.com/metabrainz/picard/pull/998, right?

mineo commented:
I think so, yes.

phw commented on plugins/addrelease/addrelease.py:
```diff
@@ -116,38 +116,84 @@ class AddClusterAsRelease(AddObjectAsEntity):
+                # disc numbers need to be changed to accommodate that.
+                self.discnumber_shift = max(self.discnumber_shift, 0 - m)
+            m = m + self.discnumber_shift
+        except Exception as e:
```
Wouldn't it be better for check for `ValueError` here?

mineo replied:
Yes.

phw commented on plugins/addrelease/addrelease.py:
```diff
@@ -116,38 +116,84 @@ class AddClusterAsRelease(AddObjectAsEntity):
+        try:
+            discnumber = metadata.get("discnumber", "1")
+            # Split off any totaldiscs information
+            discnumber = discnumber.split("/", 1)[0]
```
`discnumber` could be `None`. I think it's better to check this here instead of relying on catch-all-exception handling

mineo replied:
I don't see how `discnumber` could be None (or even an empty string). The default value passed to metadata.get is `"1"`, which should be returned in case of any falsy or None value in the tag.

phw replied:
Right, I missed the default value in `get`

phw commented on .travis.yml:
```diff
@@ -1,6 +1,10 @@
   - "3.5"
   - "3.6"
+before_install:
+  - pip3 install picard
```
As this currently breaks the CI can we do without this for now?

mineo replied:
Sure.

phw commented on plugins/addrelease/addrelease.py:
```diff
@@ -116,38 +116,84 @@ class AddClusterAsRelease(AddObjectAsEntity):
+        1
+        >>> m["discnumber"] = "1/4"
+        >>> a.extract_discnumber(m)
+        3
```
I don't get this case. Why should `a.extract_discnumber("1/4") == 3` be true? Wouldn't it be `0` (first disc)?

mineo replied:
No. The important part is that `a` has not been reset or reassigned after extracting a disc number of -2, so disc 1 is in fact disc 4 (-2, -1, 0, 1), but medium numbers start with 0, so 3. This behaviour was originally added in https://github.com/metabrainz/picard/commit/d38104821654198f5b2a2f512b2d37a4b7b6ff63 as a fix for https://tickets.metabrainz.org/browse/PICARD-407.

phw replied:
Ok, got it

phw reviewed (changes requested):
I agree that doctests is a reasonable and easy solution for having tests for the plugins, especially for the simple cases most plugins are.

Apart from some minor details it's mostly the failing build that is a problem. Maybe we should disable this for now until `pip install picard` works.

mineo committed:
Disable doctests for addrelease.py

They require being able to install & import picard from PyPI, but that doesn't
work at the moment (https://tickets.metabrainz.org/browse/PICARD-1373).

mineo committed:
addrelease: Specifically catch ValueError

The old comment mentioned a specific error, but the except clause caught all
exceptions. Be more specific during exception handling because we really don't
know if the behaviour for others makes sense.

phw commented on plugins/addrelease/addrelease.py:
```diff
@@ -171,11 +171,10 @@ def extract_discnumber(self, metadata):
             m = m + self.discnumber_shift
-        except Exception as e:
+        except ValueError as e:
             # The most likely reason for an exception at this point is a
```
I think with the updated text the "a" at the end of this line is too much ("...is because the disc number in  the tags was not a number").

phw reviewed (approved):
Apart from a small grammatical correction in the comment I think this is good to merge.

We can re-enable the doctests once `pip install picard` works, which hopefully it will with the next release and https://github.com/metabrainz/picard/pull/998 included.

mineo committed:
Remove a superfluous 'a'
